### ISSUE TITLE
Channels 2 not persisting session data set in `connect`

### ISSUE THREAD

**Issue by typistX** [django/channels#1096]
**- Your OS and runtime environment, and browser if applicable**
Development. Running a Django application from Pycharm.

**- The versions of Channels, Daphne, Django, Twisted, and your ASGI backend (channels_redis normally)**

channels==2.1.2
channels-redis==2.2.1
daphne==2.2.0
Django==1.11.6

**- What you expected to happen vs. what actually happened**

I've upgraded to Channels 2 specifically for the ability to **access and modify** the session from within a consumer, but that doesn't seem to be the case. Basically, I want to identify `AnonymousUser`s and send them messages (each one, not all of them together)

Here's my `routing.py` file:

```python
application = ProtocolTypeRouter({
    # (http->django views is added by default)
    'websocket': AuthMiddlewareStack(
        URLRouter(
            core.routing.websocket_urlpatterns
        )
    ),
})
```

Here's my `consumers.py` file:

```python
class ChatConsumer(WebsocketConsumer):
    def connect(self):
        if self.scope['user'].is_authenticated:
            user_id = str(self.scope['user'].id)
            self.scope['session']['user_identifier']= user_id
            self.group_name = user_id
        else:
            user_id = str(self.scope['user']) + str(uuid.uuid4())
            self.scope['session']['user_identifier'] = user_id
            self.group_name = user_id

        self.scope['session'].save()

        print(f" in consumer: {self.scope['session']['user_identifier']}")

        # Join room group
        async_to_sync(self.channel_layer.group_add)(
            self.group_name,
            self.channel_name
        )

        self.accept()
```

And in the view where I need to send, I'm trying to get the group_name (`user_identifier`) from the session in case of an AnonymousUser:

```python
def get_spotify_link(request):
    if request.user.is_authenticated:
        user_identifier = str(request.user.id)
    else:
        user_identifier = request.session['user_identifier']
        print(f"in get_spotify_link: {user_identifier}")
```

However, I'm intermittently (95% of the time) getting `KeyError: 'user_identifier'`.

**- How you're running Channels (runserver? daphne/runworker? Nginx/Apache in front?)**
runserver

- Console logs and full tracebacks of any errors

```
Internal Server Error: /get_spotify_link/
Traceback (most recent call last):
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/myusername/PycharmProjects/artist_notify/core/views.py", line 442, in get_spotify_link
    user_identifier = request.session['user_identifier']
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/contrib/sessions/backends/base.py", line 57, in __getitem__
    return self._session[key]
KeyError: 'user_identifier'
[2018/07/08 20:54:03] WebSocket HANDSHAKING / [127.0.0.1:65478]
 in consumer: AnonymousUser8883761c-f673-413e-82e2-413d1cbb17e3
[2018/07/08 20:54:03] WebSocket CONNECT / [127.0.0.1:65478]
Internal Server Error: /get_spotify_link/
Traceback (most recent call last):
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/myusername/PycharmProjects/artist_notify/core/views.py", line 442, in get_spotify_link
    user_identifier = request.session['user_identifier']
  File "/Users/myusername/.virtualenvs/bap_dev/lib/python3.6/site-packages/django/contrib/sessions/backends/base.py", line 57, in __getitem__
    return self._session[key]
KeyError: 'user_identifier'
```

**Comment by andrewgodwin** [comment]
So it sounds like you have the session being read from a different view than the consumer that's saving it, right? What session store are you using, and are you absolutely sure the view is called after the session is saved?

**Comment by typistX** [comment]
> So it sounds like you have the session being read from a different view than the consumer that's saving it, right?

Correct.

> What session store are you using, and are you absolutely sure the view is called after the session is saved?

`SESSION_ENGINE = 'django.contrib.sessions.backends.cached_db'`

I want to be cautious here and say that I'm 99% sure, here's what's happening:

1. User hits the "index" page (a view)
2. He connects to the websocket above - which is supposed to save the session
3. A certain click will trigger an AJAX call to a different view (this is `get_spotify_link`) - where it fails.

So yeah, the 2nd view is loaded after the session is saved. Or supposed to.

**Comment by typistX** [comment]
Actually, there's a simpler test I did:

    def connect(self):
        if self.scope['user'].is_authenticated:
            #...
        else:
            print(f"BEFORE: {self.scope['session'].get('UMI')} is the session UI variable")
            if not self.scope['session'].get('UMI', False):
                user_id = str(self.scope['user']) + str(uuid.uuid4())
                self.scope['session']['UMI'] = user_id
                self.scope['session'].modified = True # tired with and without
                self.group_name = user_id
            else:
                self.group_name = self.scope['session']['UMI']

            print(f"{self.scope['session']['UMI']} is the session UI variable")

        self.scope['session'].save()

Basically, If I refresh, and the consumer was indeed successful in saving the session, then on refresh the value of `print(f"BEFORE: {self.scope['session'].get('UMI')} is the session UI variable")` shouldn't be `None` (because they key would be found in the session). Am I correct? Then the consumer is definitely failing to save the session.

**Comment by andrewgodwin** [comment]
Yes, what you are describing is it failing to save into the session, which is impressive considering you literally call the `save()` method on that session.

Have you verified that your session backend is actually being written to when you call `save()`? Given you said it was a cached DB one I'd be worried that there might be some weird async/db transaction interaction I haven't seen before.

**Comment by typistX** [comment]
@andrewgodwin Yes. In fact, I wrote a middleware exactly to bypass this abnormality and be able to use Channels 2:

```
class SessionForAnonymousMiddleware(object):

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        print("There was *some* call")
        if request.user.is_authenticated:
            response = self.get_response(request)
            return response
        else:

            if hasattr(request, 'session') and not request.session.get('NEWUID', False):
                print("I'm being called!")
                if not request.user.is_authenticated:
                    request.session.create()
                    request.session['NEWUID'] = "AnonymousUser" + str(uuid.uuid4())
                    request.session.modified = True
                   request.session.save()
            else:
                print("no reason to call me!")
        response = self.get_response(request)
        return response
```

**Comment by andrewgodwin** [comment]
Hm, that's weird. If you could make it into a unit test that fails that would help me fix it.

**Comment by typistX** [comment]
Could this be a possible reason for my troubles? in my project root `__init__.py` I have this:

```
from __future__ import absolute_import, unicode_literals

# This will make sure the app is always imported when
# Django starts so that shared_task will use this app.
from .celery_configs import app as celery_app

__all__ = ['celery_app']
```

which imports this from `celery_configs.py`:

```
from __future__ import absolute_import, unicode_literals
import os
from celery import Celery

# set the default Django settings module for the 'celery' program.
if os.environ.get('DJANGO_SETTINGS_MODULE') == 'news_analytics.settings':
    pass
else:
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'news_analytics.settings')

app = Celery('news_analytics')


# Using a string here means the worker doesn't have to serialize
# the configuration object to child processes.
# - namespace='CELERY' means all celery-related configuration keys
#   should have a `CELERY_` prefix.
app.config_from_object('django.conf:settings', namespace='CELERY')

# Load task modules from all registered Django app configs.
app.autodiscover_tasks()


@app.task(bind=True)
def debug_task(self):
    print('Request: {0!r}'.format(self.request))
```

Maybe the fact I'm wrapping my app with `Celery` is causing all the trouble?

**Comment by andrewgodwin** [comment]
Maybe that is - try removing it and see what happens?

**Comment by andrewgodwin** [comment]
Closing due to inactivity - please reopen if the problem is not solved.

**Comment by nrempel** [comment]
Hi @andrewgodwin I'm also seeing the same problem. I'm not using celery at all.

I've tried the default session storage as well as this 3rd party driver: https://github.com/martinrusev/django-redis-sessions

It appears that the session cookie is not being set in the browser when `scope["session"].save()` is called.

```python

class ChatConsumer(WebsocketConsumer):
    def connect(self):
        self.room_name = self.scope["url_route"]["kwargs"]["room_name"]
        self.room_group_name = "chat_%s" % self.room_name

        LOGGER.info(self.scope["session"].items())
        self.scope["session"]["test"] = "set"
        LOGGER.info(self.scope["session"].items())
        self.scope["session"].save()
        LOGGER.info(self.scope["session"].session_key)

        # Join room group
        async_to_sync(self.channel_layer.group_add)(
            self.room_group_name, self.channel_name
        )

        self.accept()
```

Then refreshing my page which makes the websocket connection twice results in the following logs:


```
[0] django_1    | 2020-04-26 17:24:34,710 turnips.consumers INFO     dict_items([])
[0] django_1    | 2020-04-26 17:24:34,711 turnips.consumers INFO     dict_items([('test', 'set')])
[0] django_1    | 2020-04-26 17:24:34,722 turnips.consumers INFO     gfxw8ljci68qajcidptig15ay5txi7d6

[0] django_1    | 2020-04-26 17:24:39,054 turnips.consumers INFO     dict_items([])
[0] django_1    | 2020-04-26 17:24:39,054 turnips.consumers INFO     dict_items([('test', 'set')])
[0] django_1    | 2020-04-26 17:24:39,058 turnips.consumers INFO     bukqwz3tvm60jj8uoyli5ncgppvnoocr
```

No cookies are set in the browser.

Is there something missing here? Happy to open new ticket if this appears to be a different issue.

Thanks.

**Comment by MartenBE** [comment]
I am also experiencing this. My django right now only has a websocket endpoint (no views) and the session_key changes during every connect:

```
class GameConsumer(WebsocketConsumer):
    def connect(self):
        self.game_name = self.scope["url_route"]["kwargs"]["game_name"]
        self.session = self.scope["session"]
        self.session.save() # Necessary to generate the session_key
        logger.debug(f"GameConsumer started with session ID: {self.session.session_key}")
```

This issue should be reopened @andrewgodwin

**Comment by bschwarz67** [comment]
I also have experienced this issue. Same goes for the auth, logging in a User in channels does log them in for the correlate in the regular views. My workaround for this when the consumer sends back a message to the websocket, I immediately  use javascript to redirect to a view to do the authentication before I continue with anything else. This seems insecure/ inefficient. Below is how I'm logging in:
`sessionObject = self.scope['session']
newPlayerUsername = text_data_json['possibleNewPlayer']
newPlayer = Player(username=newPlayerUsername)
newPlayer.save()
async_to_sync(login)(self.scope, newPlayer)
self.scope['session']['test'] = 'test'
self.scope['session'].save()
print(self.scope['session']['test'])
async_to_sync(self.channel_layer.group_discard)(
      self.player_group,
      self.channel_name
)
self.player_name = self.scope['user']
self.player_group = '%s_group' % self.player_name`

**Comment by bschwarz67** [comment]
I also have experienced this issue. Same goes for the auth, logging in a User in channels does log them in for the correlate in the regular views. My workaround for this when the consumer sends back a message to the websocket, I immediately use javascript to redirect to a view to do the authentication before I continue with anything else. This seems insecure/ inefficient. Below is how I'm logging in:

```
sessionObject = self.scope['session']
newPlayerUsername = text_data_json['possibleNewPlayer']
newPlayer = Player(username=newPlayerUsername)
newPlayer.save()
async_to_sync(login)(self.scope, newPlayer)
self.scope['session']['test'] = 'test'
self.scope['session'].save()
async_to_sync(self.channel_layer.group_discard)(
       self.player_group,
       self.channel_name
)
self.player_name = self.scope['user']
self.player_group = '%s_group' % self.player_name
```
